File: packages/app/src/services/software/definitions/winscp.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'winscp';
	public name = 'WinSCP';
	public category = [
		SoftwareCategory.Network,
		SoftwareCategory.Utility,
	];
	public downloadName = 'WinSCP-Setup.exe';
	public isArchive = false;
	public shouldCacheUrl = false;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'winscp.png';
	public homepage = 'https://winscp.net';

	public async resolveDownloadUrl() {
		const downloadPageUrl = await this.getDownloadPageUrl();
		if (!downloadPageUrl) {
			return err(DownloadUrlResolveError.Generic);
		}

		const res = await fetch(downloadPageUrl);
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const text               = await res.text();
		const downloadUrlPattern = /href="(https:\/\/cdn\.winscp\.net\/files\/WinSCP-.*-Setup\.exe\?secure=.*,\d{10,})"/;
		const match              = text.match(downloadUrlPattern);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[1]);
	}

	private async getDownloadPageUrl() {
		const res = await fetch('https://winscp.net/eng/download.php');
		if (!res.ok) {
			return null;
		}

		const html  = await res.text();
		const match = html.match(/\/download\/WinSCP-.*-Setup\.exe\/download/);
		if (!match) {
			return null;
		}

		return `https://winscp.net${match[0]}`;
	}
}

File: src/lib/software/winscp.py
from re import compile
from PySide6.QtCore import Slot, QUrl
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class WinScp(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'winscp'
        self.name = 'WinSCP'
        self.category = [SoftwareCategory.Network, SoftwareCategory.Utility]
        self.download_name = 'WinSCP-Setup.exe'
        self.icon = 'winscp.png'
        self.homepage = 'https://winscp.net'

        self._initial_url = QUrl('https://winscp.net/eng/download.php')

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        html = reply.readAll().data().decode()

        if reply.url() == self._initial_url:
            download_page_pattern = compile(r'/download/WinSCP-.*-Setup\.exe/download')
            match = download_page_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                req = QNetworkRequest(f'https://winscp.net{match.group(0)}')
                self.manager.get(req)
        else:
            download_url_pattern = compile(r'href="(https://cdn\.winscp\.net/files/WinSCP-.*-Setup\.exe\?secure=.*,\d{10,})"')
            match = download_url_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                self.url_resolved.emit(match.group(1))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self._initial_url))
